src/Service/AddBulletinToQueue.php

```diff
@@ -330,10 +333,42 @@ private function buildXml() {
    */
   private function dedupeQueue() {
     if ($this->flag_dedupe === TRUE) {
-      // @todo Add the guts of this to dedupe based on $this->queue_uid.
-      // https://www.drupal.org/project/govdelivery_bulletins/issues/3096347
+      $queue = $this->getQueue();
+      $item_count = $queue->numberOfItems();
+      $removed_count = 0;
+      // The queue will keep grabbing the same item after it is released, so
+      // we need to grab them all and mass release them.
+      $queued_bulletins_to_release = [];
+      for ($i = 0; $i < $item_count; $i++) {
+        // Get a queued item
+        $queued_bulletin = $queue->claimItem(0);
+        if ($queued_bulletin) {
+          try {
+            // Process it
+            // Check if it has the same queue id as the one we will be adding.
+            if ($this->queueUid === $queued_bulletin->data->qid) {
+              // It is a match to the new one, delete it.
+              $queue->deleteItem($queued_bulletin);
+              $removed_count++;
+            }
+            else {
+              // It is not a duplicate, so release it back into the queue.
+              $queued_bulletins_to_release[$queued_bulletin->item_id] = $queued_bulletin;
+            }
+          } catch (SuspendQueueException $e) {
+            // If there was an Exception thrown because of an error
+            // Releases the item that the worker could not process.
+            // Another worker can come and process it
+            $queued_bulletins_to_release[$queued_bulletin->item_id] = $queued_bulletin;
+            continue;
```

Comment:
Is this right, @swirtSJW ? I've never seen a `continue` in a try/catch statement.

Reply:
TIL